### PR TITLE
Theme Sheet: Use information from WordPress.org for themes on Jetpack sites

### PR DESCRIPTION
**Pull request by ockham**
Allows us to find out if a theme on a Jetpack site is present on WP.org and act accordingly.

Implements https://github.com/Automattic/wp-calypso/pull/9869#issuecomment-265479960:

1. Show the tags in the features, but unlinked? ✅ 
2. Remove download as it would imply "download from Jetpack site" and doesn't seem entirely accurate to me. _-- I've kept it for now since I thought it might still be useful -- my counter-example being a Jetpress site that was recently transferred and is using a WPCOM theme the user _might_ also want to use elsewhere. Oh well, a bit contrived, I guess_ 😄 
3. You might also like: could [s]tay, but let's get without it for now. ✅ 
4. Support...
   * If it's a theme on .org, link to the theme's specific support forum (example: https://wordpress.org/support/theme/twentysixteen )
   * if it's not a theme on .org (i.e. custom, premium, etc.), hide the Support link

(Question about item 4. -- I'm not sure I've implemented it correctly, I'm currently _not_ removing the entire 'Support' tab but only the link to the theme forum. I wasn't sure I understood the idea correctly, and I thought the other two links might still make some sense even on a JP site. _Clarification needed_ 😄)

To do all that, this PR:
* Adds a `fetchThemeInformation` method to `lib/wporg`
* Changes `requestTheme` to hit that endpoint if `siteId === 'wporg'`, thus putting themes fetched from there in a new `wporg` subtree
* Adds a normalization util to account for different attribute names in objects obtained from the WP.org API
* Changes the `getTheme` selector so that if a theme on a JP site is requested, supplementary data (upon availability) such as demo URI, download URI, or `taxonomies` is merged into the returned object.
* Adds an `isWporgTheme( state, themeId )` aliasing `!! getTheme( state, 'wpcom', themeId )`
* Modifies the theme sheet to conditionally hide some UI elements (such as the 'Open Live Demo' link) if some given theme attributes aren't available (for cases where no data from WP.org was available).

To test:
* Try the theme sheet on a Jetpack site, ideally for a theme that's not on WPCOM. Try two different themes, one that _is_  on WP.org, and one that isn't.
* WP.org theme: Check that the following UI elements are present and work properly:
  * 'Open Live Demo' link
  * 'Download' button
  * Theme Features (but they don't link anywhere!)
  * Related theme suggestions are _not_ there
  * 'Visit forum' link in the 'Support' tab should point to theme-specific forum on WP.org
* non-WP.org theme: None of the above UI elements should be present
* Also check that wpcom theme sheets still work as before.

(For reference, here's an example response containing all the fields we can get from WP.org: https://api.wordpress.org/themes/info/1.1/?action=theme_information&request%5Bslug%5D=twentyfourteen)

### CODE REVIEW

File: client/components/data/query-active-theme/index.jsx [hunk 1]
```diff
@@ -12,10 +12,7 @@ import { isRequestingActiveTheme } from 'state/themes/selectors';
 
 class QueryActiveTheme extends Component {
 	static propTypes = {
-		siteId: PropTypes.oneOfType( [
-			PropTypes.number,
-			PropTypes.oneOf( [ 'wpcom' ] )
```
Comment (ockham) [inline client/components/data/query-active-theme/index.jsx]: A theme can only be active on an actual site, not on all of WP.com 😄 

(Discovered while adding `wporg` to some other `propTypes` and `schema`s.)

**Comment by seear** [comment]
We may want to hide the _Features_ block if it is empty:

screen shot 2016-12-09 at 13 24 36

File: client/my-sites/theme/main.jsx [hunk 2]
```diff
@@ -451,6 +469,7 @@ const ThemeSheet = React.createClass( {
 		return (
 			<Main className="theme__sheet">
 				<QueryTheme themeId={ this.props.id } siteId={ siteIdOrWpcom } />
+				{ isJetpack && <QueryTheme themeId={ this.props.id } siteId="wporg" /> }
```
Comment (seear) [inline client/my-sites/theme/main.jsx]: Do we need the two instances of `<QueryTheme/>`?

File: client/state/themes/test/actions.js [hunk 3]
```diff
@@ -229,19 +229,19 @@ describe( 'actions', () => {
...
-				.reply( 404, {
-					error: 'unknown_theme',
-					message: 'Unknown theme'
-				} );
-		} );
-
 		context( 'with a wpcom site', () => {
+			useNock( ( nock ) => {
+				nock( 'https://public-api.wordpress.com:443' )
+					.persist()
```
Comment (seear) [inline client/state/themes/test/actions.js]: I don't think these `persist` calls are generally necessary unless you want the same result multiple times.

**Comment by ockham** [comment]
> Code wise looks good, although hard to follow: to much stuff is going on at the same time. 👍

Yeah, sorry about that. I guess I could've split it into even more PRs, but then again, I already did that twice, so I wanted to get it done...

File: client/my-sites/theme/main.jsx (hunk 2 above)
Comment (ockham) [inline client/my-sites/theme/main.jsx]: Yeah, the first one fetches the theme object from the JP site, and the second one fetches the supplementary one from the WP.org API.

File: client/state/themes/test/actions.js (hunk 3 above)
Comment (ockham) [inline client/state/themes/test/actions.js]: Breaks tests if I remove it for all `wpcom`, `wporg`, and JP 😕

File: client/my-sites/theme/main.jsx [hunk 4]
```diff
@@ -334,7 +346,13 @@ const ThemeSheet = React.createClass( {
 	},
 
 	renderDownload() {
-		if ( isPremium( this.props ) ) {
+		// Don't render download button:
+		// * If it's a premium theme
+		// * If it's on a Jetpack site, and the theme object doesn't have a 'download' attr
+		//   Note that not having a 'download' attr would be permissible for a theme on WPCOM
+		//   since we don't provide any for some themes found on WordPress.org (notably the 'Twenties').
```
Comment (ockham) [inline client/my-sites/theme/main.jsx]: I checked a bunch of other Twenties to which the same seems to apply, but possibly not all of them.

**Comment by ockham** [comment]
> We may want to hide the Features block if it is empty

Done in 8cbee7e

**Comment by ockham** [comment]
(Mini-fix to `schema.js` to get rid of a state validation warning.)

File: client/my-sites/theme/main.jsx [hunk 5]
```diff
@@ -310,16 +320,22 @@ const ThemeSheet = React.createClass( {
...
+			return null;
+		}
+
+		const themeFeatures = taxonomies.theme_feature.map( function( item ) {
 			const term = isValidTerm( item.slug ) ? item.slug : `feature:${ item.slug }`;
 			return (
 				<li key={ 'theme-features-item-' + item.slug }>
-					<a href={ `/design/filter/${ term }/${ siteSlug || '' }` }>{ item.name }</a>
+					{ isJetpack
+						? <a>{ item.name }</a>
```
Comment (omarjackman) [inline client/my-sites/theme/main.jsx]: Just curious if adding the `<a>` is actually doing anything here. Are styles applied to it because its an anchor?
Maybe it could just be `{ item.name }`

**Comment by folletto** [comment]
> Remove download as it would imply "download from Jetpack site" and doesn't seem entirely accurate to me. -- I've kept it for now since I thought it might still be useful -- my counter-example being a Jetpress site that was recently transferred and is using a WPCOM theme the user might also want to use elsewhere. Oh well, a bit contrived, I guess

Keep it only if it's already working flawlessly. If you need even just 10 minutes of extra work to make it work across .com/Jetpack (how does Jetpack zip the theme to allow its download?) remove it. It's not a priority right now.

> Question about item 4. -- I'm not sure I've implemented it correctly, I'm currently not removing the entire 'Support' tab but only the link to the theme forum. I wasn't sure I understood the idea correctly, and I thought the other two links might still make some sense even on a JP site. Clarification needed 😄

Ok, let me branch it out!  :)

For Jetpack:

1. If it's a local / uploaded / custom theme → the Support tab disappears as we can't provide support from .com, nor we know if there's a .org forum.
2. If it's a Business User → show the tab with only `renderContactUsCard` ("Need extra help?")

**Comment by ockham** [comment]
> Keep it only if it's already working flawlessly. If you need even just 10 minutes of extra work to make it work across .com/Jetpack (how does Jetpack zip the theme to allow its download?) remove it. It's not a priority right now.

Oh, maybe I wasn't quite clear -- we're only showing the Download link if the theme is found on WP.org, and has a download URI (pointing to a ready-made zip, that is) already. Otherwise we're hiding it.

**Comment by folletto** [comment]
> Oh, maybe I wasn't quite clear -- we're only showing the Download link if the theme is found on WP.org, and has a download URI (pointing to a ready-made zip, that is) already. Otherwise we're hiding it.

Ok cool, I thought the whole "check .org's availability" was postponed. :)

Excellent then! :D

**Comment by ockham** [comment]
> For Jetpack:

> Ok cool, I thought the whole "check .org's availability" was postponed. :)

Hehe, no, that's what this PR does 😄 (cf. https://github.com/Automattic/wp-calypso/pull/9869#issuecomment-265565566):

> I thought initially that the is-on-dotorg check might be a bit too difficult to implement given that we're on a rather tight schedule, but some tinkering (#9875) revealed that it's probably not that hard after all.

**Comment by ockham** [comment]
> If it's a local / uploaded / custom theme → the Support tab disappears as we can't provide support from .com, nor we know if there's a .org forum.

Oh, I was trying to implement https://github.com/Automattic/wp-calypso/pull/9869#issuecomment-265483782

> My suggestion would be:

> - If it's a theme on .org, link to the theme's specific support forum (example: https://wordpress.org/support/theme/twentysixteen )

So if you look at the 'Support' tab's forum link for a theme on a JP site that's also found on WP.org, it will point to that theme's WP.org support forum right now.

**Comment by folletto** [comment]
> So if you look at the 'Support' tab's forum link for a theme on a JP site that's also found on WP.org, it will point to that theme's WP.org support forum right now.

Same as the above. I was working on the assumption .org's checking wasn't included right now, in that case:

1. If there is a .org forum, show its block.
2. If there is Business Plan enabled, show its block.
3. If the Support tab is empty (none of the above) hide it.
